### PR TITLE
Bug 1784582: Added Helm CLI download links

### PR DESCRIPTION
**Pull request by pedjak**
This PR adds to OpenShift Console the additional Helm CLI download link(s).

### CODE REVIEW

**Comment by pedjak** [comment]
Screenshot-20191210192800-976x545

After merging this request, the CLI download page will look as above.

**Comment by sbose78** [comment]
CC @siamaksade @bparees @soltysh

File: manifests/07-downloads-helm.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+apiVersion: console.openshift.io/v1
+kind: ConsoleCLIDownload
+metadata:
+  name: helm-download-links
+spec:
+  description: |
+    Helm helps you manage Kubernetes applications.
+    Helm Charts help you define, install, and upgrade even the most complex Kubernetes application.
```
Comment (bparees) [inline manifests/07-downloads-helm.yaml]: as discussed yesterday, this needs to indicate it is tech preview.

Comment (sbose78) [reply manifests/07-downloads-helm.yaml]: Ah right!

Comment (pedjak) [reply manifests/07-downloads-helm.yaml]: @siamaksade can you suggest a proper wording here, so that it is clear that we are talking about tech preview?

Comment (siamaksade) [reply manifests/07-downloads-helm.yaml]: @pedjak @sbose78 `Helm 3 is a package manager for Kubernetes applications which enables defining, installing and upgrading applications packaged as Helm Charts. Helm 3 is in Technology Preview feature on OpenShift 4`

Not sure if we can include links there but if we could, then _Technology Preview_ could be linked to: https://access.redhat.com/support/offerings/techpreview

**Comment by pedjak** [comment]
It looks that tests are failing due to infra issues:

```
Error 400: The number of members in the policy (1,501) is larger than the maximum allowed size 1,500., badRequest"
```

How to make them pass?

File: manifests/07-downloads-helm.yaml (hunk 1 above)
Comment (pedjak) [reply manifests/07-downloads-helm.yaml]: @siamaksade @sbose78 actually embedding html tags works, not sure if this is intended or not. 

I have updated the text, take a look how the text is now rendered.
Screenshot-20191213001833-1340x141

File: manifests/07-downloads-helm.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+apiVersion: console.openshift.io/v1
+kind: ConsoleCLIDownload
+metadata:
+  name: helm-download-links
+spec:
+  description: |
+    Helm 3 is a package manager for Kubernetes applications which enables defining,
+    installing and upgrading applications packaged as Helm Charts.
+    Helm 3 is in <a href=https://access.redhat.com/support/offerings/techpreview>Technology Preview</a> feature on OpenShift 4.
```
Comment (sbose78) [inline manifests/07-downloads-helm.yaml]: Predrag, have you verified if it shows up as a clickable URL?

Comment (pedjak) [reply manifests/07-downloads-helm.yaml]: Yes, see the screenshot attached above,

Comment (bparees) [reply manifests/07-downloads-helm.yaml]: either:
`Helm 3 is in Technology Preview on OpenShift`
or
`Helm 3 is a Technology Preview feature on Openshift`

but `Helm 3 is in Technology Preview feature on OpenShift` is not a grammatically correct sentence.

**Comment by bparees** [comment]
@sbose78 @pedjak https://mirror.openshift.com/pub/openshift-v4/helm/latest is not a valid link.  When do you plan to have the content there? I do not want us to ship ocp4 with a broken link in the console.

also i would expect the url to be something under https://mirror.openshift.com/pub/openshift-v4/clients/

**Comment by pedjak** [comment]
@bparees thanks, correct the grammar issue and changed the download link to https://mirror.openshift.com/pub/openshift-v4/clients/helm/latest My understanding is that we want to deliver helm with 4.3 release, right @sbose78 ?

**Comment by bparees** [comment]
/hold cancel
@benjaminapetersen @spadgett i think this can merge if you're ok with it, assuming we can get confirmation the helm binary will be there by the time 4.3 GAs.

**Comment by sbose78** [comment]
Yes, that's correct, Predrag & Ben. The binary would be there when 4.3 releases.

File: manifests/07-downloads-helm.yaml (hunk 1 above)
Comment (benjaminapetersen) [reply manifests/07-downloads-helm.yaml]: Evaluating the embedded link, I don't recall if that is intentional atm, but definitely need to look at this.

Comment (benjaminapetersen) [reply manifests/07-downloads-helm.yaml]: Ok this text is rendered within `SyncMarkdownView`, we should be good here.
https://github.com/openshift/console/blob/ac67e8c3c66b73cea9f38cc0deb990b64c674259/frontend/public/components/command-line-tools.tsx#L25

**Comment by benjaminapetersen** [comment]
You could also update the tests
https://github.com/openshift/console-operator/blob/master/test/e2e/downloads_test.go#L24
To prove that the link functions in 4.3/4.4.

File: manifests/07-downloads-helm.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+apiVersion: console.openshift.io/v1
+kind: ConsoleCLIDownload
+metadata:
+  name: helm-download-links
+spec:
+  description: |
+    Helm 3 is a package manager for Kubernetes applications which enables defining,
+    installing and upgrading applications packaged as Helm Charts.
+    Helm 3 is a <a href=https://access.redhat.com/support/offerings/techpreview>Technology Preview</a> feature on OpenShift 4.
```
Comment (benjaminapetersen) [inline manifests/07-downloads-helm.yaml]: Quote the `href`?

**Comment by benjaminapetersen** [comment]
We should also squash the commits before merge.

File: manifests/07-downloads-helm.yaml (hunk 1 above)
Comment (benjaminapetersen) [reply manifests/07-downloads-helm.yaml]: https://github.com/openshift/console/blob/88dc7f8343826e710e052ec559657648cd97a07a/frontend/public/components/markdown-view.tsx#L16
for reference, allowed HTML entities.

File: manifests/07-downloads-helm.yaml (hunk 2 above)
Comment (benjaminapetersen) [reply manifests/07-downloads-helm.yaml]: There is also an e2e suite in console that verifies the CLIDownload links.  You could add a few lines here as well:

https://github.com/openshift/console/blob/e3080430f4ac297619b57062410c461333e62c9f/frontend/integration-tests/tests/crd-extensions.scenario.ts#L15

**Comment by pedjak** [comment]
@benjaminapetersen I would add the check to the tests, but we still do not have the binaries ready for the download, or you mean that we should only check if the link gets rendered? BTW, Checking ODO links is also not a part of the test atm.

I have squashed the commits.

File: manifests/07-downloads-helm.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+apiVersion: console.openshift.io/v1
+kind: ConsoleCLIDownload
+metadata:
+  name: helm-download-links
+spec:
+  description: |
+    Helm 3 is a package manager for Kubernetes applications which enables defining,
+    installing and upgrading applications packaged as Helm Charts.
+    Helm 3 is a <a href="https://access.redhat.com/support/offerings/techpreview">Technology Preview</a> feature on OpenShift 4.
```
Comment (spadgett) [inline manifests/07-downloads-helm.yaml]: Prefer markdown-style links

```suggestion
    Helm 3 is a [Technology Preview](https://access.redhat.com/support/offerings/techpreview) feature on OpenShift 4.
```

File: manifests/07-downloads-helm.yaml [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+  displayName: helm - Helm 3 CLI
+  links:
+    - href: 'https://mirror.openshift.com/pub/openshift-v4/clients/helm/latest'
+      text: Download helm
```
Comment (spadgett) [inline manifests/07-downloads-helm.yaml]: Missing final newline

File: manifests/07-downloads-helm.yaml [hunk 6]
```diff
@@ -0,0 +1,13 @@
+apiVersion: console.openshift.io/v1
+kind: ConsoleCLIDownload
+metadata:
+  name: helm-download-links
+spec:
+  description: |
+    Helm 3 is a package manager for Kubernetes applications which enables defining,
+    installing and upgrading applications packaged as Helm Charts.
```
Comment (spadgett) [inline manifests/07-downloads-helm.yaml]: Oxford comma please :)

```suggestion
    installing, and upgrading applications packaged as Helm Charts.
```

**Comment by pedjak** [comment]
@spadgett - I have included the proposed fixes, thanks for your feedback. I cannot add the test at the moment because we do not have binaries ready for download. Speaking about the links, should we add actually individual DL links for each of the platforms (linux, osx, windows) or leave it as it is. I have followed ODO approach, @siamaksade what do you think?

Speaking about testing, the existing `TestDownloadsEndpoint()` checks only if `oc` can be downloaded. Do we have somewhere browser tests that check if the page is properly rendered? Do we use https://github.com/tebeka/selenium or similar?

**Comment by bparees** [comment]
fyi we are passed code freeze, this did not merge.  Barring a request for an extremely urgent exception, this will not ship in 4.3.0GA, we can consider including in in the first z-stream.

**Comment by siamaksade** [comment]
@bparees pushing it into z-stream 4.3.1 is fine. Does is need a BZ ticket?

@pedjak we should follow `odo` in 4.3. In 4.4 we will have fixed versions for each OCP release and should switch to the `oc` approach.

**Comment by bparees** [comment]
> @bparees pushing it into z-stream 4.3.1 is fine. Does is need a BZ ticket?

yes you'll need a BZ to backport/cherry-pick this PR to release-4.3.

**Comment by siamaksade** [comment]
BZ ticket created for backport: https://bugzilla.redhat.com/show_bug.cgi?id=1784582

**Comment by siamaksade** [comment]
@sbose78  @bparees can we go ahead with mering this PR to master (4.4)? to then create a new PR for 4.3 branch

**Comment by bparees** [comment]
> @sbose78 @bparees can we go ahead with mering this PR to master (4.4)? to then create a new PR for 4.3 branch

if @spadgett and @benjaminapetersen are happy they can merge it.

File: manifests/07-downloads-helm.yaml (hunk 4 above)
Comment (benjaminapetersen) [reply manifests/07-downloads-helm.yaml]: Looks like the link still needs to be updated to markdown style `Helm 3 is a [Technology Preview](https://access.redhat.com/support/offerings/techpreview) feature on OpenShift 4.`

File: manifests/07-downloads-helm.yaml (hunk 6 above)
Comment (benjaminapetersen) [reply manifests/07-downloads-helm.yaml]: I think this comment was missed as well.

File: manifests/07-downloads-helm.yaml (hunk 4 above)
Comment (pedjak) [reply manifests/07-downloads-helm.yaml]: done.

**Comment by pedjak** [comment]
@benjaminapetersen, I have  pushed hopefully all fixes.

**Comment by benjaminapetersen** [comment]
I passed you the bug for this:  https://bugzilla.redhat.com/show_bug.cgi?id=1784582

**Comment by bparees** [comment]
/hold

please look into these failures.
https://prow.svc.ci.openshift.org/job-history/origin-ci-test/pr-logs/directory/pull-ci-openshift-console-operator-master-e2e-aws-console

**Comment by pedjak** [comment]
@bparees the test were passing the last time [here](https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_console-operator/362/pull-ci-openshift-console-operator-master-e2e-aws-console/660) but the only big difference to the actual PR is that Tech Preview link was declared using `<a href.../>` syntax.

**Comment by benjaminapetersen** [comment]
The console e2e test suite is struggling atm, we are definitely looking into it.

**Comment by spadgett** [comment]
We're working on a fix for the console e2e tests in https://github.com/openshift/console/pull/3847

**Comment by spadgett** [comment]
/hold cancel 
/retest 
CI is fixed

**Pull request merged**